File: kvstore-5/py/kvstore/wal.py
"""Write-Ahead Log implementation for persistence."""
import json
import os
from typing import Any, Optional
from .utils import current_timestamp


class WAL:
    """Write-ahead log for persisting operations."""

    def __init__(self, data_dir: str):
        """Initialize WAL.

        Args:
            data_dir: Directory to store WAL file
        """
        self.data_dir = data_dir
        self.wal_path = os.path.join(data_dir, 'wal.log')

        # Create data directory if it doesn't exist
        os.makedirs(data_dir, exist_ok=True)

        # Create WAL file if it doesn't exist
        if not os.path.exists(self.wal_path):
            open(self.wal_path, 'w').close()

    def append_set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Append a SET operation to the WAL.

        Args:
            key: The key being set
            value: The value being stored
            ttl: Time-to-live in seconds (optional)
        """
        entry = {
            'op': 'set',
            'key': key,
            'value': value,
            'timestamp': current_timestamp()
        }
        if ttl is not None:
            entry['ttl'] = ttl

        with open(self.wal_path, 'a') as f:
            f.write(json.dumps(entry) + '\n')

    def append_delete(self, key: str) -> None:
        """Append a DELETE operation to the WAL.

        Args:
            key: The key being deleted
        """
        entry = {
            'op': 'delete',
            'key': key,
            'timestamp': current_timestamp()
        }

        with open(self.wal_path, 'a') as f:
            f.write(json.dumps(entry) + '\n')

    def replay(self, store) -> None:
        """Replay WAL entries to restore state.

        Args:
            store: KVStore instance to replay operations on
        """
        if not os.path.exists(self.wal_path):
            return

        with open(self.wal_path, 'r') as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue

                try:
                    entry = json.loads(line)
                    op = entry.get('op')

                    if op == 'set':
                        key = entry['key']
                        value = entry['value']
                        ttl = entry.get('ttl')

                        # Calculate remaining TTL if original operation had TTL
                        if ttl is not None:
                            timestamp = entry.get('timestamp', current_timestamp())
                            elapsed = current_timestamp() - timestamp
                            remaining_ttl = ttl - elapsed

                            # Only set if not expired
                            if remaining_ttl > 0:
                                store.set(key, value, remaining_ttl)
                        else:
                            store.set(key, value, None)

                    elif op == 'delete':
                        key = entry['key']
                        store.delete(key)

                except (json.JSONDecodeError, KeyError):
                    # Skip malformed entries
                    continue

File: kvstore-5/typescript/src/src/wal.ts
/**
 * Write-Ahead Log for persistence
 */

import * as fs from 'fs/promises';
import * as fsSync from 'fs';
import * as path from 'path';
import { KVStore } from './store';
import { getCurrentTimestamp } from './utils';

interface WALOperation {
  op: 'set' | 'delete';
  key: string;
  value?: any;
  ttl?: number;
  timestamp: number;
}

export class WAL {
  private walPath: string;
  private writeStream: fsSync.WriteStream | null = null;

  constructor(private dataDir: string) {
    this.walPath = path.join(dataDir, 'wal.log');
  }

  /**
   * Initialize WAL: create directory and open write stream
   */
  async initialize(): Promise<void> {
    // Create data directory if it doesn't exist
    await fs.mkdir(this.dataDir, { recursive: true });

    // Ensure WAL file exists
    try {
      await fs.access(this.walPath);
    } catch {
      // Create empty file if it doesn't exist
      await fs.writeFile(this.walPath, '');
    }

    // Open write stream in append mode
    this.writeStream = fsSync.createWriteStream(this.walPath, { flags: 'a' });
  }

  /**
   * Log a set operation
   */
  async logSet(key: string, value: any, ttl?: number): Promise<void> {
    const operation: WALOperation = {
      op: 'set',
      key,
      value,
      timestamp: getCurrentTimestamp(),
    };

    if (ttl !== undefined) {
      operation.ttl = ttl;
    }

    await this.writeOperation(operation);
  }

  /**
   * Log a delete operation
   */
  async logDelete(key: string): Promise<void> {
    const operation: WALOperation = {
      op: 'delete',
      key,
      timestamp: getCurrentTimestamp(),
    };

    await this.writeOperation(operation);
  }

  /**
   * Write an operation to the WAL
   */
  private async writeOperation(operation: WALOperation): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.writeStream) {
        reject(new Error('WAL not initialized'));
        return;
      }

      const line = JSON.stringify(operation) + '\n';
      this.writeStream.write(line, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }

  /**
   * Replay WAL to restore state
   */
  async replay(store: KVStore): Promise<void> {
    try {
      // Check if WAL file exists
      await fs.access(this.walPath);
    } catch {
      // WAL doesn't exist yet, nothing to replay
      return;
    }

    const content = await fs.readFile(this.walPath, 'utf-8');
    const lines = content.split('\n').filter((line) => line.trim());

    for (const line of lines) {
      try {
        const operation: WALOperation = JSON.parse(line);

        if (operation.op === 'set') {
          // Calculate remaining TTL if present
          let remainingTTL: number | undefined = undefined;
          if (operation.ttl !== undefined) {
            const elapsed = getCurrentTimestamp() - operation.timestamp;
            remainingTTL = operation.ttl - elapsed;

            // Skip if already expired
            if (remainingTTL <= 0) {
              continue;
            }
          }

          store.set(operation.key, operation.value, remainingTTL);
        } else if (operation.op === 'delete') {
          store.delete(operation.key);
        }
      } catch (err) {
        console.error('Error replaying WAL operation:', err);
      }
    }
  }

  /**
   * Close the WAL
   */
  async close(): Promise<void> {
    return new Promise((resolve) => {
      if (this.writeStream) {
        this.writeStream.end(() => {
          resolve();
        });
      } else {
        resolve();
      }
    });
  }
}
